topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,511 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
+	metadata["Speed"] = int64(intf.LinkSpeed)
+	metadata["State"] = interfaceUpDown(intf.AdminUpDown)
+	metadata["Duplex"] = interfaceDuplex(intf.LinkDuplex)
+	metadata["VrfID"] = vrfID
+
+	if strings.HasPrefix(name, "host-") {
+		metadata["InterfaceName"] = name[5:]
+	}
+
 	var err error
-	node := p.Ctx.Graph.LookupFirstNode(metadata)
+	node := p.getInterface(intf.SwIfIndex)
 	if node == nil {
 		node, err = p.Ctx.Graph.NewNode(graph.GenID(), metadata)
 		if err != nil {
 			p.Ctx.Logger.Error(err)
 			return nil
 		}
+		p.NotifyEvent(graph.NodeAdded, node)
+
 		if _, err = p.Ctx.Graph.Link(p.vppRootNode, node, topology.OwnershipMetadata()); err != nil {
 			p.Ctx.Logger.Error(err)
 			return nil
 		}
+	} else {
+		p.Ctx.Graph.SetMetadata(node, metadata)
+		p.NotifyEvent(graph.NodeUpdated, node)
 	}
-
-	tr := p.Ctx.Graph.StartMetadataTransaction(node)
-	defer tr.Commit()
-	tr.AddMetadata("Driver", "vpp")
-	tr.AddMetadata("Name", strings.Trim(string(intf.InterfaceName), "\000"))
-	tr.AddMetadata("IfIndex", int64(intf.SwIfIndex))
-	tr.AddMetadata("MAC", interfaceMAC(intf.L2Address[:intf.L2AddressLength]))
-	tr.AddMetadata("MTU", int64(intf.LinkMtu))
-	tr.AddMetadata("Speed", int64(intf.LinkSpeed))
-	state := interfaceUpDown(intf.AdminUpDown)
-	tr.AddMetadata("State", state)
-	if state != "DOWN" {
-		tr.AddMetadata("Duplex", interfaceDuplex(intf.LinkDuplex))
-	}
-	tr.AddMetadata("VrfID", p.getInterfaceVrfID(ch, intf.SwIfIndex))
-
 	return node
 }
 
 func interfaceNeedUpdate(i1, i2 *interfaces.SwInterfaceDetails) bool {
-	if (i1.LinkMtu != i2.LinkMtu) ||
-		(i1.LinkSpeed != i2.LinkSpeed) ||
-		(i1.LinkDuplex != i2.LinkDuplex) {
-		return true
-	}
-	return false
+	return i1.LinkMtu != i2.LinkMtu ||
+		i1.LinkSpeed != i2.LinkSpeed ||
+		i1.LinkDuplex != i2.LinkDuplex
 }
 
-func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) {
-	p.createOrUpdateInterface(ch, intf)
+func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails, metadata graph.Metadata) *graph.Node {
+	return p.createOrUpdateInterface(ch, intf, metadata)
 }
 
 func (p *Probe) eventDelInterface(node *graph.Node) {
-	p.Ctx.Graph.Lock()
-	defer p.Ctx.Graph.Unlock()
 	if err := p.Ctx.Graph.DelNode(node); err != nil {
+		p.NotifyEvent(graph.NodeDeleted, node)
 		p.Ctx.Logger.Error(err)
 	}
 }
 
 func (p *Probe) interfaceEventsEnableDisable(ch api.Channel, enable bool) {
-	ed := uint32(0)
+	req := &interfaces.WantInterfaceEvents{PID: uint32(os.Getpid())}
 	if enable {
-		ed = uint32(1)
-	}
-	req := &interfaces.WantInterfaceEvents{
-		EnableDisable: ed,
-		PID:           uint32(os.Getpid()),
+		req.EnableDisable = 1
 	}
 	msg := &interfaces.WantInterfaceEventsReply{}
-	err := ch.SendRequest(req).ReceiveReply(msg)
-	if err != nil {
+	if err := ch.SendRequest(req).ReceiveReply(msg); err != nil {
 		p.Ctx.Logger.Error(err)
 	}
 }
 
-func (p *Probe) interfacesEvents() {
-	defer p.wg.Done()
+type memifSocket struct {
+	*memif.MemifDetails
```

Comment:
This is an example where interface configuration is defined be directly re-using structures from the VPP binary API. Since that changes quite often, it might not compile if skydive is build against a different VPP version.